pkg/engine/lifecycletest/taint_test.go

```diff
@@ -0,0 +1,358 @@
+	assert.False(t, finalResource.Taint, "taint should be cleared after replacement")
+	assert.False(t, finalResource.Delete, "resource should not be marked for deletion")
+}
+
+// TestTaintNoChanges tests that a tainted resource forces replacement even when there are no other changes.
+func TestTaintNoChanges(t *testing.T) {
+	t.Parallel()
+
+	replaceCount := 0
+	loaders := []*deploytest.ProviderLoader{
+		deploytest.NewProviderLoader("pkgA", semver.MustParse("1.0.0"), func() (plugin.Provider, error) {
+			return &deploytest.Provider{
+				CreateF: func(_ context.Context, req plugin.CreateRequest) (plugin.CreateResponse, error) {
+					return plugin.CreateResponse{
+						ID:         resource.ID("id-" + string(rune('0'+replaceCount))),
+						Properties: req.Properties,
+						Status:     resource.StatusOK,
+					}, nil
+				},
+				DeleteF: func(_ context.Context, req plugin.DeleteRequest) (plugin.DeleteResponse, error) {
+					replaceCount++
+					return plugin.DeleteResponse{Status: resource.StatusOK}, nil
+				},
+				DiffF: func(_ context.Context, req plugin.DiffRequest) (plugin.DiffResponse, error) {
+					// No changes in properties
+					return plugin.DiffResponse{
+						Changes: plugin.DiffNone,
+					}, nil
+				},
+			}, nil
+		}),
+	}
+
+	programF := deploytest.NewLanguageRuntimeF(func(_ plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+		// Register the same resource with same properties
+		_, err := monitor.RegisterResource("pkgA:m:typA", "resA", true, deploytest.ResourceOptions{
+			Inputs: resource.PropertyMap{
+				"foo": resource.NewStringProperty("bar"),
+			},
+		})
+		return err
+	})
+
+	hostF := deploytest.NewPluginHostF(nil, nil, programF, loaders...)
+
+	p := &lt.TestPlan{
+		Options: lt.TestUpdateOptions{T: t, HostF: hostF, SkipDisplayTests: true},
+	}
+
+	// Run initial update
+	project := p.GetProject()
+	snap, err := lt.TestOp(Update).Run(project, p.GetTarget(t, nil), p.Options, false, p.BackendClient, nil)
+	require.NoError(t, err)
+
+	// Run update without taint - should be no-op
+	snap, err = lt.TestOp(Update).Run(project, p.GetTarget(t, snap), p.Options, false, p.BackendClient, nil)
+	require.NoError(t, err)
+	assert.Equal(t, 0, replaceCount, "no replacement should occur without taint")
+
+	// Taint the resource
+	for _, r := range snap.Resources {
+		if r.URN.Name() == "resA" {
+			r.Taint = true
+			break
+		}
+	}
+
+	// Run update with taint - should force replacement
+	snap, err = lt.TestOp(Update).Run(project, p.GetTarget(t, snap), p.Options, false, p.BackendClient, nil)
+	require.NoError(t, err)
+	assert.Equal(t, 1, replaceCount, "replacement should occur due to taint")
+
+	// Verify taint is cleared
+	for _, r := range snap.Resources {
+		if r.URN.Name() == "resA" {
+			assert.False(t, r.Taint, "taint should be cleared after replacement")
+			assert.Equal(t, resource.ID("id-0"), r.ID, "resource should have new ID")
+			break
```

Comment:
Why are we breaking out of the loop here?  We should make sure there's only one resource with name `resA` in the snapshot left.

Also, we should assert that we found a resource with name `resA`.